src/Compilers/CSharp/Portable/Binder/SwitchBinder.cs

```diff
@@ -27,90 +29,102 @@ protected SwitchBinder(Binder next, SwitchStatementSyntax switchSyntax)
 
         internal static SwitchBinder Create(Binder next, SwitchStatementSyntax switchSyntax)
         {
-            // When pattern-matching is enabled, we use a completely different binder and binding
-            // strategy for switch statements. Once we have confirmed that it is totally upward
-            // compatible with the existing syntax and semantics, we will remove *this* binder
-            // and use the new one for binding all switch statements.
             var parseOptions = switchSyntax?.SyntaxTree?.Options as CSharpParseOptions;
             return
-                (parseOptions?.IsFeatureEnabled(MessageID.IDS_FeaturePatternMatching) != false &&
-                (parseOptions?.Features.ContainsKey("typeswitch") != false || IsPatternSwitch(switchSyntax)))
+                // in C# 6 and earlier, we use the old binder. In C# 7 and later, we use the new binder which
+                // is capable of binding both the old and new syntax. However, the new binder does not yet
+                // lead to a translation that fully supports edit-and-continue.
+                (parseOptions?.IsFeatureEnabled(MessageID.IDS_FeaturePatternMatching) != false || parseOptions?.Features.ContainsKey("typeswitch") != false)
                 ? new PatternSwitchBinder(next, switchSyntax)
                 : new SwitchBinder(next, switchSyntax);
         }
 
-        private static bool IsPatternSwitch(SwitchStatementSyntax switchSyntax)
+
+        protected bool PatternsEnabled =>
+            ((CSharpParseOptions)this._switchSyntax.SyntaxTree.Options)?.IsFeatureEnabled(MessageID.IDS_FeaturePatternMatching) != false;
+
+        protected BoundExpression SwitchGoverningExpression
         {
-            foreach (var section in switchSyntax.Sections)
+            get
             {
-                if (section.Labels.Any(SyntaxKind.CasePatternSwitchLabel))
+                if (_switchGoverningExpression == null)
                 {
-                    return true;
+                    var switchGoverningDiagnostics = new DiagnosticBag();
+                    var boundSwitchExpression = BindSwitchExpression(_switchSyntax.Expression, switchGoverningDiagnostics);
+                    _switchGoverningDiagnostics = switchGoverningDiagnostics;
```

Comment:
It creates a memory barrier, allowing the data to be seen across threads.

---
In reply to: [70331887](https://github.com/dotnet/roslyn/pull/12416#discussion_r70331887) [](ancestors = 70331887)